src/query/api/v1/handler/namespace/update.go

```diff
@@ -0,0 +1,221 @@
+}
+
+func (h *UpdateHandler) ServeHTTP(
+	svc handleroptions.ServiceNameAndDefaults,
+	w http.ResponseWriter,
+	r *http.Request,
+) {
+	ctx := r.Context()
+	logger := logging.WithContext(ctx, h.instrumentOpts)
+
+	md, rErr := h.parseRequest(r)
+	if rErr != nil {
+		logger.Error("unable to parse request", zap.Error(rErr))
+		xhttp.Error(w, rErr.Inner(), rErr.Code())
+		return
+	}
+
+	opts := handleroptions.NewServiceOptions(svc, r.Header, nil)
+	nsRegistry, err := h.Update(md, opts)
+	if err != nil {
+		logger.Error("unable to update namespace", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	resp := &admin.NamespaceGetResponse{
+		Registry: &nsRegistry,
+	}
+
+	xhttp.WriteProtoMsgJSONResponse(w, resp, logger)
+}
+
+func (h *UpdateHandler) parseRequest(r *http.Request) (*admin.NamespaceUpdateRequest, *xhttp.ParseError) {
+	defer r.Body.Close()
+	rBody, err := xhttp.DurationToNanosBytes(r.Body)
+	if err != nil {
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
+
+	updateReq := new(admin.NamespaceUpdateRequest)
+	if err := jsonpb.Unmarshal(bytes.NewReader(rBody), updateReq); err != nil {
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
+
+	if err := validateUpdateRequest(updateReq); err != nil {
+		err := fmt.Errorf("unable to validate update request: %w", err)
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
+
+	return updateReq, nil
+}
+
+// Ensure that only fields we allow to be updated (e.g. retention period) are
+// non-zero. Uses reflection to be resilient against adding more immutable
+// fields to namespaceOptions but forgetting to validate them here.
+func validateUpdateRequest(req *admin.NamespaceUpdateRequest) error {
+	if req.Name == "" {
+		return errEmptyNamespaceName
+	}
+
+	if req.Options == nil {
+		return errEmptyNamespaceOptions
+	}
+
+	if req.Options.RetentionOptions == nil {
+		return errEmptyRetentionOptions
+	}
+
+	optsVal := reflect.ValueOf(*req.Options)
+
+	for i := 0; i < optsVal.NumField(); i++ {
+		field := optsVal.Field(i)
+		fieldName := optsVal.Type().Field(i).Name
+		if !field.IsZero() && fieldName != fieldNameRetentionOptions {
+			return fmt.Errorf("%s: %w", fieldName, errNamespaceFieldImmutable)
+		}
+	}
+
+	optsVal = reflect.ValueOf(*req.Options.RetentionOptions)
+	for i := 0; i < optsVal.NumField(); i++ {
+		field := optsVal.Field(i)
+		fieldName := optsVal.Type().Field(i).Name
+		if !field.IsZero() && fieldName != fieldNameRetetionPeriod {
+			return fmt.Errorf("%s.%s: %w", fieldNameRetentionOptions, fieldName, errNamespaceFieldImmutable)
+		}
+	}
+
+	return nil
+}
+
+// Update updates a namespace.
+func (h *UpdateHandler) Update(
+	updateReq *admin.NamespaceUpdateRequest,
+	opts handleroptions.ServiceOptions,
+) (nsproto.Registry, error) {
+	var emptyReg = nsproto.Registry{}
+
+	kvOpts := kv.NewOverrideOptions().
+		SetEnvironment(opts.ServiceEnvironment).
+		SetZone(opts.ServiceZone)
+
+	store, err := h.client.Store(kvOpts)
```

Comment:
Can we put this on the "handleroptions.ServiceOptions"?

Something like:
```go
// KVOverrideOptions constructs the KV override options from the options.
func (opts *ServiceOptions) KVOverrideOptions() kv.OverrideOptions {
	return kv.NewOverrideOptions().
		SetEnvironment(opts.ServiceEnvironment).
		SetZone(opts.ServiceZone)
}
```

Here's an existing one:
```go
// ServiceID constructs a cluster services ID from the options.
func (opts *ServiceOptions) ServiceID() services.ServiceID {
	return services.NewServiceID().
		SetName(opts.ServiceName).
		SetEnvironment(opts.ServiceEnvironment).
		SetZone(opts.ServiceZone)
}
```